src/include/btree_inline.h

```diff
@@ -766,6 +766,21 @@ __wt_tree_modify_set(WT_SESSION_IMPL *session)
         /* Assert we never dirty a checkpoint handle. */
         WT_ASSERT(session, !WT_READING_CHECKPOINT(session));
 
+        /*
+         * We should never set a btree dirty when checkpoint is triggered by RTS, recovery or when
+         * closing the connection. Those specific scenarios should always leave the database clean.
+         * The only exception is related to the metadata file: it is expected to be marked as dirty
+         * whenever a btree is checkpointed.
+         */
+        if (WT_SESSION_BTREE_SYNC(session) && !WT_IS_METADATA(session->dhandle) &&
+          !FLD_ISSET(S2C(session)->timing_stress_flags, WT_TIMING_STRESS_CHECKPOINT_EVICT_PAGE)) {
+            WT_ASSERT_ALWAYS(session, !F_ISSET(session, WT_SESSION_ROLLBACK_TO_STABLE), "%s",
+              "A btree is marked dirty during RTS");
+            WT_ASSERT_ALWAYS(session, !F_ISSET(S2C(session), WT_CONN_RECOVERING), "%s",
+              "A btree is marked dirty during recovery");
+            WT_ASSERT_ALWAYS(session, !F_ISSET(S2C(session), WT_CONN_CLOSING), "%s",
```

Comment:
The close flag must be `WT_CONN_CLOSING_CHECKPOINT`.

Reply:
Thank you, edited in 5676471d4e523b4f2109700cc09389e73b495f75

Reply:
After thinking about this again, I think it is better to keep `WT_CONN_CLOSING` as it covers a bigger period of time and we don't want to set a btree dirty during this entire time, not only during the ckpt.

Reply:
No. The `WT_CONN_CLOSING` is covering the end part of the shutdown after the shutdown checkpoint.

Reply:
I see, my bad. Just noticed that we actually never clear `WT_CONN_CLOSING_CHECKPOINT`.